src/wayland/dmabuf/mod.rs

```diff
@@ -11,471 +11,470 @@
-    log: ::slog::Logger,
+/// Gets the contents of a [`Dmabuf`] backed [`Buffer`].
+///
+/// If the buffer is managed by the dmabuf handler, the [`Dmabuf`] is returned.
+///
+/// If the buffer is not managed by the dmabuf handler (whether the buffer is a different kind of buffer,
+/// such as an shm buffer or is not managed by smithay), this function will return an [`UnmanagedResource`]
+/// error.
+pub fn get_dmabuf(buffer: &Buffer) -> Result<Dmabuf, UnmanagedResource> {
+    Ok(buffer.buffer_data::<Dmabuf>()?.ok_or(UnmanagedResource)?.clone())
 }
 
-impl<H> ParamsHandler<H>
-where
-    H: for<'a> FnMut(&Dmabuf, DispatchData<'a>) -> bool + 'static,
-{
-    fn add(
-        &mut self,
-        params: &BufferParams,
-        fd: RawFd,
-        plane_idx: u32,
-        offset: u32,
-        stride: u32,
-        modifier: u64,
-    ) {
-        // protocol checks:
-        // Cannot reuse a params:
-        if self.used {
-            params.as_ref().post_error(
-                ParamError::AlreadyUsed as u32,
-                "This buffer_params has already been used to create a buffer.".into(),
-            );
-            return;
-        }
-        // plane_idx is not too large
-        if plane_idx >= self.max_planes {
-            // plane_idx starts at 0
-            params.as_ref().post_error(
-                ParamError::PlaneIdx as u32,
-                format!("Plane index {} is out of bounds.", plane_idx),
-            );
-            return;
-        }
-        // plane_idx has already been set
-        if self.pending_planes.iter().any(|d| d.plane_idx == plane_idx) {
-            params.as_ref().post_error(
-                ParamError::PlaneSet as u32,
-                format!("Plane index {} is already set.", plane_idx),
-            );
-            return;
+/// Macro to delegate implementation of the linux dmabuf to [`DmabufState`].
+///
+/// You must also implement [`DmabufHandler`] to use this.
+#[macro_export]
+macro_rules! delegate_dmabuf {
+    ($ty: ty) => {
+        type __ZwpLinuxDmabufV1 =
+            $crate::reexports::wayland_protocols::unstable::linux_dmabuf::v1::server::zwp_linux_dmabuf_v1::ZwpLinuxDmabufV1;
+        type __ZwpLinuxBufferParamsV1 =
+            $crate::reexports::wayland_protocols::unstable::linux_dmabuf::v1::server::zwp_linux_buffer_params_v1::ZwpLinuxBufferParamsV1;
+
+        $crate::reexports::wayland_server::delegate_global_dispatch!($ty: [
+            __ZwpLinuxDmabufV1
+        ] => $crate::wayland::dmabuf::DmabufState);
+
+        $crate::reexports::wayland_server::delegate_dispatch!($ty: [
+            __ZwpLinuxDmabufV1,
+            __ZwpLinuxBufferParamsV1
+        ] => $crate::wayland::dmabuf::DmabufState);
+    };
+}
+
+const GLOBAL_VERSION: u32 = 3;
+
+impl DmabufParamsData {
+    /// Emits a protocol error if the params have already been used to create a dmabuf.
+    ///
+    /// This returns true if the protocol object has already been used.
+    fn if_used_then_err(
```

Comment:
This does not seem to be a very explicit name for what it does when seeing it a call site.

What about something like `ensure_unused`, which returns `true` if it is indeed not yet used?